src/main/scala/com/fulcrumgenomics/bam/pileup/PileupBuilder.scala

```diff
@@ -0,0 +1,159 @@
+
+  /** Checks to see if all SAM record filters accept the record. */
+  protected def acceptRecord(rec: SamRecord): Boolean = quickAcceptRecord(rec) && recFilters.forall(fn => fn(rec))
+
+  /** Checks to see if all pileup entry filters accept the pileup entry. */
+  protected def acceptEntry(entry: PileupEntry): Boolean = quickAcceptEntry(entry) && entryFilters.forall(fn => fn(entry))
+
+  /** Constructs a pileup, at the single requested position, from the records returned by the iterator.
+    *
+    * @param recs the collection of coordinate-sorted SAM records that may or may not overlap the position
+    * @param refName the name of the reference sequence on which the position resides
+    * @param pos the 1-based position on the reference sequence at which to construct the pileup
+    */
+  def build(recs: IterableOnce[SamRecord], refName: String, pos: Int): Pileup[PileupEntry] = {
+    val refIndex = dict(refName).index.ensuring(_ >= 0, s"Unknown reference sequence name: $refName")
+    val pile     = IndexedSeq.newBuilder[PileupEntry]
+    if (recs.knownSize >= 0) pile.sizeHint(pile.knownSize)
+
+    @inline def testAndAdd(entry: PileupEntry): Unit = if (this.acceptEntry(entry)) pile.addOne(entry)
+
+    recs.iterator.bufferBetter
+      .dropWhile(rec => rec.refIndex < refIndex)
+      .takeWhile(rec => rec.refIndex == refIndex && rec.start <= pos + 1)
+      .filter { rec =>
+        lazy val startsWithInsertion = PileupBuilder.startsWithInsertion(rec)
+        var compare: Boolean = !rec.unmapped
+        if (compare) compare = this.acceptRecord(rec)
+        if (compare) compare = rec.end >= pos
+        if (compare) compare = rec.start <= pos || startsWithInsertion
+        if (compare) compare = if (excludeMapPositionOutsideFrInsert && rec.isFrPair) {
+          PileupBuilder.positionInsideFrInsert(rec, refIndex = refIndex, pos = pos).contains(true)
+        } else { compare }
+        compare
+      }
+      .foreach { rec =>
+        if (rec.start == pos + 1) { // This site must be an insertion in the read that is at the start of the read.
+          testAndAdd(InsertionEntry(rec, 0))
+        } else {
+          val offset = rec.readPosAtRefPos(pos, returnLastBaseIfDeleted = false)
+          if (offset == 0) { // This site must be deleted within the read.
+            val deletionPosition = rec.readPosAtRefPos(pos, returnLastBaseIfDeleted = true)
+            testAndAdd(DeletionEntry(rec, deletionPosition - 1))
+          } else { // This site must be a matched site within the read. Also check and add read-end insertions.
+            testAndAdd(BaseEntry(rec, offset - 1))
+            if (offset < rec.length - 1 && rec.refPosAtReadPos(offset + 1) == 0) testAndAdd(InsertionEntry(rec, offset))
+          }
+        }
+      }
+
+    Pileup(refName = refName, refIndex = refIndex, pos = pos, pile = pile.result())
+  }
+}
+
+/** Companion object for [[PileupBuilder]]. */
+object PileupBuilder {
```

Comment:
I've seen them both ways in `fgbio`. I'll move it! Although I find them better at the bottom since I'm almost always first interested in the classes and traits and companion objects IMHO are just that... companions.